src/python/pants/rules/core/strip_source_roots.py

```diff
@@ -82,11 +83,9 @@ def determine_source_root(self, *, source_root_config: SourceRootConfig) -> str:
   if target_adaptor.type_alias == Files.alias():
     return SourceRootStrippedSources(snapshot=target_adaptor.sources.snapshot)
 
-  # NB: The sentinel file does not need to actually exist. It is solely used to determine what the
-  # source root is for the target.
-  sentinel_file = PurePath(hydrated_target.address.spec_path, "SENTINEL").as_posix()
+  build_file = PurePath(hydrated_target.address.spec_path, "BUILD").as_posix()
```

Comment:
Why not just use `hydrated_target.address.spec_path`? Why add any suffix at all?

Reply:
This line is so that we behave the same as before, where we end up stripping `hydrated_target.address.spec_path` as the prefix when there is no source root. 

Without it, we would end up stripping `Path(hydrated_target.address.spec_path).parent` because the rule `strip_source_roots_from_snapshot` always calls `Path(representative_path).parent` to determine the source root when there is none found and `--source-unmatched=create`.

For example, if we pass the representative path as `not-a-real-source-root/util` for the target `not-a-real-source-root/util:strutil`, then we would try stripping the prefix `not-a-real-source-root` from every file, rather than `not-a-real-source-root/util` like we previously did.

Reply:
This is pretty unintuitive. The reader will wonder "why a BUILD file, specifically?" 

Why not change that behavior in `strip_source_roots_from_snapshot`? It was cargo-culted from v1 code that assumed that the thing being passed in was a file, and the logic was "if there is no source root, then this file's parent is its source root".  But we can assume it's a dir and return that, since we control the inputs here.
 
And in fact, in a follow up change, I think we can just always have the `--source-unmatched=error` behavior (and not register the --source-unmatched option in v2-only mode). The `create` behavior was legacy behavior that doesn't actually make much sense. Everything you specifically request source-root-stripping for should have a source root. It should be an error otherwise.

Reply:
I don't know how to make it more intuitive, given that using `SENTINEL` also is problematic and we don't want the perf hit of using `await Get[BuildFileAddress](Address)` to get the actual build file path.

Instead, I'm going to stick with this but am very happy to deprecate `--source-unmatched=create`, meaning that this code will be removed sooner than later. In that followup, I'll add a `TODO` to remove this part of the code once the deprecation is complete.